### DISCUSSION TITLE
edl as fastboot

### DISCUSSION THREAD

**Discussion by electriquo** [bkerler/edl#586] [Q&A]
Firstly, I **really want to thank @bkerler for all this amazing work**.

While flashing [Xiaomi Mi A3](https://www.gsmarena.com/xiaomi_mi_a3-9769.php) phone, the phone shutdown and the computer crashed, and now the phone does not turn on and as a result I cannot get the phone boot into fastboot mode in order to re-flash it.

Digging the internet led me here and I have few questions:

1. Can edl be used to flash stock firmware to the phone?
2. Does edl has equivalent fastboot commands?
3. Is there any reference/example/guide for converting fastboot commands to edl commands (I would really appreciate any help)? E.g. The `flash_all.sh` file that is bundled with the stock firmware looks has these fastboot commands
    ```shell
    fastboot flash crclist images/crclist.txt
    fastboot flash sparsecrclist images/sparsecrclist.txt
    fastboot flash tz_a images/tz.mbn
    fastboot flash xbl_a images/xbl.elf
    fastboot flash xbl_config_a images/xbl_config.elf
    fastboot flash rpm_a images/rpm.mbn
    fastboot flash abl_a images/abl.elf
    fastboot flash tz_b images/tz.mbn
    fastboot flash xbl_b images/xbl.elf
    fastboot flash xbl_config_b images/xbl_config.elf
    fastboot flash rpm_b images/rpm.mbn
    fastboot flash abl_b images/abl.elf
    fastboot flash devcfg_a images/devcfg.mbn
    fastboot flash hyp_a images/hyp.mbn
    fastboot flash cmnlib_a images/cmnlib.mbn
    fastboot flash cmnlib64_a images/cmnlib64.mbn
    fastboot flash keymaster_a images/km4.mbn
    fastboot flash devcfg_b images/devcfg.mbn
    fastboot flash hyp_b images/hyp.mbn
    fastboot flash cmnlib_b images/cmnlib.mbn
    fastboot flash cmnlib64_b images/cmnlib64.mbn
    fastboot flash keymaster_b images/km4.mbn
    fastboot flash qupfw_a images/qupv3fw.elf
    fastboot flash imagefv_a images/imagefv.elf
    fastboot flash bluetooth_a images/BTFM.bin
    fastboot flash uefisecapp_a images/uefi_sec.mbn
    fastboot flash storsec_a images/storsec.mbn
    fastboot flash logfs images/logfs_ufs_8mb.bin
    fastboot flash dsp_a images/dspso.bin
    fastboot flash qupfw_b images/qupv3fw.elf
    fastboot flash imagefv_b images/imagefv.elf
    fastboot flash bluetooth_b images/BTFM.bin
    fastboot flash uefisecapp_b images/uefi_sec.mbn
    fastboot flash storsec_b images/storsec.mbn
    fastboot flash dsp_b images/dspso.bin
    fastboot erase boot_a
    fastboot erase boot_b
    fastboot erase system_a
    fastboot erase system_b
    fastboot erase vendor_a
    fastboot erase vendor_b
    fastboot erase mdtp_a
    fastboot erase mdtp_b
    fastboot erase userdata
    fastboot flash apdp images/apdp.mbn
    fastboot flash modem_a images/NON-HLOS.bin
    fastboot flash vendor_a images/vendor.img
    fastboot flash system_a images/system.img
    fastboot flash dtbo_a images/dtbo.img
    fastboot flash vbmeta_a images/vbmeta.img
    fastboot flash modem_b images/NON-HLOS.bin
    fastboot flash vendor_b images/vendor.img
    fastboot flash system_b images/system_other.img
    fastboot flash dtbo_b images/dtbo.img
    fastboot flash vbmeta_b images/vbmeta.img
    fastboot flash product_a images/product.img
    fastboot flash product_b images/product.img
    fastboot flash userdata images/userdata.img
    fastboot flash metadata images/metadata.img
    fastboot flash boot_a images/boot.img
    fastboot flash boot_b images/boot.img
    fastboot reboot
    ```

Any related information and guidance would be really appreciated since I cannot afford buying a new phone.
Thanks you.

If it helps, I do see that edl detects the phone (connecting it while edl is running) when running edl naively

```shell
$ python edl --debugmode
Keystone library is missing (optional).
Qualcomm Sahara / Firehose Client V3.62 (c) B.Kerler 2018-2024.
main - Trying with no loader given ...
main - Waiting for the device
DeviceClass
DeviceClass - [LIB]: Couldn't detect the device. Is it connected ?
.DeviceClass
DeviceClass - [LIB]: Couldn't detect the device. Is it connected ?
.main - Device detected :)
DeviceClass
DeviceClass - [LIB]: Timed out
DeviceClass
DeviceClass - [LIB]: Timed out
DeviceClass
DeviceClass - [LIB]: Timed out
DeviceClass
DeviceClass - [LIB]: Timed out
DeviceClass
DeviceClass - [LIB]: Timed out
DeviceClass
DeviceClass - [LIB]: Timed out
DeviceClass
DeviceClass - [LIB]: Timed out
DeviceClass
DeviceClass - [LIB]: Timed out
DeviceClass
DeviceClass - [LIB]: Timed out
DeviceClass
DeviceClass - [LIB]: Timed out
DeviceClass
DeviceClass - [LIB]: [Errno 60] Operation timed out
DeviceClass
DeviceClass - [LIB]: [Errno 60] Operation timed out
^C
```

**Comment by ferouzkassim** [comment]
the device is not connected make sure qualcomm 9008 drivers are
installed correctly and then run again you can try the loaders that come
along the repo or try the loader that comes along the firmware you
downloaded

On Mon, Oct 7, 2024 at 3:12 PM electriquo ***@***.***> wrote:

> Firstly, I *really want to thank @bkerler <https://github.com/bkerler>
> for all this amazing work*.
>
> While flashing Xiaomi Mi A3
> <https://www.gsmarena.com/xiaomi_mi_a3-9769.php> phone, the phone
> shutdown and the computer crashed, and now the phone does not turn on and
> as a result I cannot get the phone boot into fastboot mode in order to
> re-flash it.
>
> Digging the internet led me here and I have few questions:
>
>    1. Can edl be used to flash stock firmware to the phone?
>    2. Does edl has equivalent fastboot commands?
>    3. Is there any reference/example/guide for converting fastboot
>    commands to edl commands (I would really appreciate any help)? E.g. The
>    flash_all.sh file that is bundled with the stock firmware looks has
>    these fastboot commands
>
>    fastboot flash crclist images/crclist.txt
>    fastboot flash sparsecrclist images/sparsecrclist.txt
>    fastboot flash tz_a images/tz.mbn
>    fastboot flash xbl_a images/xbl.elf
>    fastboot flash xbl_config_a images/xbl_config.elf
>    fastboot flash rpm_a images/rpm.mbn
>    fastboot flash abl_a images/abl.elf
>    fastboot flash tz_b images/tz.mbn
>    fastboot flash xbl_b images/xbl.elf
>    fastboot flash xbl_config_b images/xbl_config.elf
>    fastboot flash rpm_b images/rpm.mbn
>    fastboot flash abl_b images/abl.elf
>    fastboot flash devcfg_a images/devcfg.mbn
>    fastboot flash hyp_a images/hyp.mbn
>    fastboot flash cmnlib_a images/cmnlib.mbn
>    fastboot flash cmnlib64_a images/cmnlib64.mbn
>    fastboot flash keymaster_a images/km4.mbn
>    fastboot flash devcfg_b images/devcfg.mbn
>    fastboot flash hyp_b images/hyp.mbn
>    fastboot flash cmnlib_b images/cmnlib.mbn
>    fastboot flash cmnlib64_b images/cmnlib64.mbn
>    fastboot flash keymaster_b images/km4.mbn
>    fastboot flash qupfw_a images/qupv3fw.elf
>    fastboot flash imagefv_a images/imagefv.elf
>    fastboot flash bluetooth_a images/BTFM.bin
>    fastboot flash uefisecapp_a images/uefi_sec.mbn
>    fastboot flash storsec_a images/storsec.mbn
>    fastboot flash logfs images/logfs_ufs_8mb.bin
>    fastboot flash dsp_a images/dspso.bin
>    fastboot flash qupfw_b images/qupv3fw.elf
>    fastboot flash imagefv_b images/imagefv.elf
>    fastboot flash bluetooth_b images/BTFM.bin
>    fastboot flash uefisecapp_b images/uefi_sec.mbn
>    fastboot flash storsec_b images/storsec.mbn
>    fastboot flash dsp_b images/dspso.bin
>    fastboot erase boot_a
>    fastboot erase boot_b
>    fastboot erase system_a
>    fastboot erase system_b
>    fastboot erase vendor_a
>    fastboot erase vendor_b
>    fastboot erase mdtp_a
>    fastboot erase mdtp_b
>    fastboot erase userdata
>    fastboot flash apdp images/apdp.mbn
>    fastboot flash modem_a images/NON-HLOS.bin
>    fastboot flash vendor_a images/vendor.img
>    fastboot flash system_a images/system.img
>    fastboot flash dtbo_a images/dtbo.img
>    fastboot flash vbmeta_a images/vbmeta.img
>    fastboot flash modem_b images/NON-HLOS.bin
>    fastboot flash vendor_b images/vendor.img
>    fastboot flash system_b images/system_other.img
>    fastboot flash dtbo_b images/dtbo.img
>    fastboot flash vbmeta_b images/vbmeta.img
>    fastboot flash product_a images/product.img
>    fastboot flash product_b images/product.img
>    fastboot flash userdata images/userdata.img
>    fastboot flash metadata images/metadata.img
>    fastboot flash boot_a images/boot.img
>    fastboot flash boot_b images/boot.img
>    fastboot reboot
>
>
> Any related information and guidance would be really appreciated since I
> cannot afford buying a new phone.
> Thanks you.
>
> If it helps, I do see that edl detects the phone (connecting it while edl
> is running) when running edl naively
>
> $ python edl --debugmode
> Keystone library is missing (optional).
> Qualcomm Sahara / Firehose Client V3.62 (c) B.Kerler 2018-2024.
> main - Trying with no loader given ...
> main - Waiting for the device
> DeviceClass
> DeviceClass - [LIB]: Couldn't detect the device. Is it connected ?.DeviceClassDeviceClass - [LIB]: Couldn't detect the device. Is it connected ?
> .main - Device detected :)
> DeviceClass
> DeviceClass - [LIB]: Timed out
> DeviceClass
> DeviceClass - [LIB]: Timed out
> DeviceClass
> DeviceClass - [LIB]: Timed out
> DeviceClass
> DeviceClass - [LIB]: Timed out
> DeviceClass
> DeviceClass - [LIB]: Timed out
> DeviceClass
> DeviceClass - [LIB]: Timed out
> DeviceClass
> DeviceClass - [LIB]: Timed out
> DeviceClass
> DeviceClass - [LIB]: Timed out
> DeviceClass
> DeviceClass - [LIB]: Timed out
> DeviceClass
> DeviceClass - [LIB]: Timed out
> DeviceClass
> DeviceClass - [LIB]: [Errno 60] Operation timed out
> DeviceClass
> DeviceClass - [LIB]: [Errno 60] Operation timed out
> ^C
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bkerler/edl/discussions/586>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFRDI3FVUIJA2JUP7WZM7MDZ2J3DRAVCNFSM6AAAAABPPZAJ5KVHI2DSMVQWIX3LMV43ERDJONRXK43TNFXW4OZXGI4DKNRUGY>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
YOURS
FEROUZ KASSIM
TEAM DROID
THANK YOU

**Reply by electriquo** [reply]
Hi @ferouzkassim 

> the device is not connected make sure qualcomm 9008 drivers are installed correctly 

I am running on macOS, are there any drivers?

> you can try the loaders that come along the repo

```shell
$ ls -lR Loaders | grep 9008
```

From the phone technical specification, it's written that it uses Qualcomm SDM665 Snapdragon 665 but it does not appear in the [loaders](https://github.com/bkerler/Loaders) or maybe i just do not know how to find it among all the loaders. Any lead?

> try the loader that comes along the firmware you downloaded

I have the stock firmware, but I see no loader there. Any idea?

Besides, do you have any lead about my original questions (e.g. converting from fastboot to edl commands?)

**Comment by ferouzkassim** [comment]
to be honest am not sure about he compatibility with macos but the linux
and windows operating was reported supported however also you should look
at the firmware you downloaded usually they have an extension of .mbn for
the programmer firehose (loader)

On Mon, Oct 7, 2024 at 3:30 PM electriquo ***@***.***> wrote:

> Hi @ferouzkassim <https://github.com/ferouzkassim>
>
> the device is not connected make sure qualcomm 9008 drivers are installed
> correctly
>
> I am running on macOS, are there any drivers?
>
> you can try the loaders that come along the repo
>
> From the phone technical specification, it's written that it uses Qualcomm
> SDM665 Snapdragon 665 but it does not appear in the loaders
> <https://github.com/bkerler/Loaders> or maybe i just do not know how to
> find it among all the loaders. Any lead?
>
> try the loader that comes along the firmware you downloaded
>
> I have the stock firmware, but I see no loader there. Any idea?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bkerler/edl/discussions/586#discussioncomment-10867058>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFRDI3BBDNTCZPST3PPSL5LZ2J5FLAVCNFSM6AAAAABPPZAJ5KVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTAOBWG4YDKOA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
YOURS
FEROUZ KASSIM
TEAM DROID
THANK YOU

**Reply by electriquo** [reply]
have too many `.mbn` files

```shell
$ find Xiaomi_Mi_A3_V10.3.4.0.PFQMIXM_20190815.0000.00_Global_9.0_XFT -name '*.mbn'
Xiaomi_Mi_A3_V10.3.4.0.PFQMIXM_20190815.0000.00_Global_9.0_XFT/Firmware/images/km4.mbn
Xiaomi_Mi_A3_V10.3.4.0.PFQMIXM_20190815.0000.00_Global_9.0_XFT/Firmware/images/rpm.mbn
Xiaomi_Mi_A3_V10.3.4.0.PFQMIXM_20190815.0000.00_Global_9.0_XFT/Firmware/images/hyp.mbn
Xiaomi_Mi_A3_V10.3.4.0.PFQMIXM_20190815.0000.00_Global_9.0_XFT/Firmware/images/apdp.mbn
Xiaomi_Mi_A3_V10.3.4.0.PFQMIXM_20190815.0000.00_Global_9.0_XFT/Firmware/images/uefi_sec.mbn
Xiaomi_Mi_A3_V10.3.4.0.PFQMIXM_20190815.0000.00_Global_9.0_XFT/Firmware/images/cmnlib.mbn
Xiaomi_Mi_A3_V10.3.4.0.PFQMIXM_20190815.0000.00_Global_9.0_XFT/Firmware/images/devcfg.mbn
Xiaomi_Mi_A3_V10.3.4.0.PFQMIXM_20190815.0000.00_Global_9.0_XFT/Firmware/images/storsec.mbn
Xiaomi_Mi_A3_V10.3.4.0.PFQMIXM_20190815.0000.00_Global_9.0_XFT/Firmware/images/tz.mbn
Xiaomi_Mi_A3_V10.3.4.0.PFQMIXM_20190815.0000.00_Global_9.0_XFT/Firmware/images/cmnlib64.mbn
```

Assume I will figure the device connection, is there any way to convert `fastboot` to `edl` commands?

**Comment by vmikhnevych** [comment]
If the device is detected, but you get this:
DeviceClass - [LIB]: Timed out

Most likely you should use sudo to run edl commands. 
And use a correct --memory parameter (UFS or EMMC, whichever your phone has).
